pkg/commands/loadgen.go

```diff
@@ -0,0 +1,214 @@
+		Default("1000").IntVar(&loadgenCommand.activeSeries)
+	cmd.Flag("scrape-interval", "period to send metrics").
+		Default("15s").DurationVar(&loadgenCommand.scrapeInterval)
+	cmd.Flag("parallelism", "how many metrics to send simultaneously").
+		Default("10").IntVar(&loadgenCommand.parallelism)
+	cmd.Flag("batch-size", "how big a batch to send").
+		Default("100").IntVar(&loadgenCommand.batchSize)
+	cmd.Flag("write-timeout", "timeout for write requests").
+		Default("500ms").DurationVar(&loadgenCommand.writeTimeout)
+
+	cmd.Flag("query-url", "").
+		Required().StringVar(&loadgenCommand.queryURL)
+	cmd.Flag("query", "query to run").
+		Default("sum(node_cpu_seconds_total)").StringVar(&loadgenCommand.query)
+	cmd.Flag("query-parallelism", "number of queries to run in parallel").
+		Default("10").IntVar(&loadgenCommand.queryParallelism)
+	cmd.Flag("query-timeout", "").
+		Default("20s").DurationVar(&loadgenCommand.queryTimeout)
+	cmd.Flag("query-duration", "length of query").
+		Default("1h").DurationVar(&loadgenCommand.queryDuration)
+
+	cmd.Flag("metrics-listen-address", "address to serve metrics on").
+		Default(":8080").StringVar(&loadgenCommand.metricsListenAddress)
+}
+
+func (c *LoadgenCommand) run(k *kingpin.ParseContext) error {
+	writeURL, err := url.Parse(c.writeURL)
+	if err != nil {
+		return err
+	}
+
+	writeClient, err := remote.NewClient(0, &remote.ClientConfig{
+		URL:     &config.URL{URL: writeURL},
+		Timeout: model.Duration(c.writeTimeout),
+	})
+	if err != nil {
+		return err
+	}
+	c.writeClient = writeClient
+
+	queryClient, err := api.NewClient(api.Config{
+		Address: c.queryURL,
+	})
+	if err != nil {
+		return err
+	}
+	c.queryClient = v1.NewAPI(queryClient)
+
+	http.Handle("/metrics", promhttp.Handler())
+	go http.ListenAndServe(c.metricsListenAddress, nil)
+
+	c.wg.Add(c.parallelism)
+	c.wg.Add(c.queryParallelism)
+
+	metricsPerShard := c.activeSeries / c.parallelism
+	for i := 0; i < c.activeSeries; i += metricsPerShard {
+		go c.runWriteShard(i, i+metricsPerShard)
+	}
+
+	for i := 0; i < c.queryParallelism; i++ {
+		go c.runQueryShard()
+	}
+
+	c.wg.Wait()
+	return nil
+}
+
+func (c *LoadgenCommand) runWriteShard(from, to int) {
+	defer c.wg.Done()
+	ticker := time.NewTicker(c.scrapeInterval)
+	c.runScrape(from, to)
+	for range ticker.C {
+		c.runScrape(from, to)
+	}
+}
+
+func (c *LoadgenCommand) runScrape(from, to int) {
+	for i := from; i < to; i += c.batchSize {
+		if err := c.runBatch(i, i+c.batchSize); err != nil {
+			log.Printf("error sending batch: %v", err)
+		}
+	}
+	fmt.Printf("sent %d samples\n", to-from)
+}
+
+func (c *LoadgenCommand) runBatch(from, to int) error {
+	var (
+		req = prompb.WriteRequest{
+			Timeseries: make([]prompb.TimeSeries, 0, to-from),
+		}
+		now = time.Now().UnixNano() / int64(time.Millisecond)
+	)
+
+	for i := from; i < to; i++ {
+		timeseries := prompb.TimeSeries{
+			Labels: []prompb.Label{
+				{Name: "__name__", Value: "node_cpu_seconds_total"},
```

Comment:
Might be better off choosing a metric name that doesn't already exist or make it configurable similar to the `-query` flag above. It's NBD to leave it as is. It's easy enough to change later.

Reply:
Good idea but yeah, lets fix later.